ca/certificate-authority.go

```diff
@@ -215,176 +217,397 @@ func (ca *CertificateAuthorityImpl) RevokeCertificate(serial string, reasonCode
-		ca.log.Audit(fmt.Sprintf("Failed RPC to store at SA, orphaning certificate: pem=[%s] err=[%v]", certPEM, err))
-		return emptyCert, err
+		ca.Log.Audit(fmt.Sprintf("Failed RPC to store at SA, orphaning certificate: pem=[%s] err=[%v]", certPEM, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
+		return
 	}
 
-	// Submit the certificate to any configured CT logs
+	// Perform post-issuance checks
 	certObj, err := x509.ParseCertificate(certDER)
 	if err != nil {
-		ca.log.Warning(fmt.Sprintf("Post-Issuance OCSP failed parsing Certificate: %s", err))
-		return cert, nil
+		ca.Log.Warning(fmt.Sprintf("Post-Issuance OCSP failed parsing Certificate: %s", err))
+		return
+	}
+
+	err = ca.MatchesCSR(cert, csr)
+	if err != nil {
+		err = core.InternalServerError(err.Error())
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		ca.Log.Audit(fmt.Sprintf("Failed RPC to store at SA, orphaning certificate: pem=[%s] err=[%v]", certPEM, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
+		return
+	}
+
+	// Here's where we finally declare success
+	// Now that the certificate has been committed, we can fill out
+	// the rest of the log event and mark it as successful
+	logEvent.SerialNumber = cert.Serial
+	logEvent.CommonName = certObj.Subject.CommonName
+	logEvent.Names = certObj.DNSNames
+	logEvent.NotBefore = certObj.NotBefore
+	logEvent.NotAfter = certObj.NotAfter
+	logEvent.ResponseTime = cert.Issued
+	logEventResult = "successful"
+	ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusValid)
+
+	// Attempt to generate the OCSP Response now. If this raises an error, it is
+	// logged but is not returned to the caller, as an error at this point does
+	// not constitute an issuance failure.
+	serial := core.SerialToString(certObj.SerialNumber)
+	ocspRequest := ocsp.SignRequest{
+		Certificate: certObj,
+		Status:      string(core.OCSPStatusGood),
+	}
+
+	ocspResponse, err := ca.OCSPSigner.Sign(ocspRequest)
+	if err != nil {
+		ca.Log.Warning(fmt.Sprintf("Post-Issuance OCSP failed signing: %s", err))
+		return
+	}
+
+	err = ca.SA.UpdateOCSP(serial, ocspResponse)
+	if err != nil {
+		ca.Log.Warning(fmt.Sprintf("Post-Issuance OCSP failed storing: %s", err))
+		return
+	}
+
+	// Submit the certificate to any configured CT logs
+	ca.Publisher.SubmitToCT(certObj.Raw)
+
+	return
+}
+
+// MatchesCSR tests the contents of a generated certificate to make sure
+// that the PublicKey, CommonName, and DNSNames match those provided in
+// the CSR that was used to generate the certificate. It also checks the
+// following fields for:
+//		* notBefore is not more than 24 hours ago
+//		* BasicConstraintsValid is true
+//		* IsCA is false
+//		* ExtKeyUsage only contains ExtKeyUsageServerAuth & ExtKeyUsageClientAuth
+//		* Subject only contains CommonName & Names
+func (ca *CertificateAuthorityImpl) MatchesCSR(cert core.Certificate, csr *x509.CertificateRequest) (err error) {
```

Comment:
Note for future reviewers: This is just copied from `ra/registration-authority.go`, as are the post-issuance checks.  The RA used to apply these checks after issuance, but with async, the RA isn't in the loop post-issuance, so the CA does the checks.